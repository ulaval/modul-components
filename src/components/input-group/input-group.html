<fieldset class="m-input-group"
          :class="classes">
    <legend v-if="legend && visible"
            class="m-input-group__legend">{{ legend }}</legend>
    <slot :disabled="isDisabled"
          :waiting="isWaiting"
          :error="hasError"
          :valid="isValid"
          :readonly="readonly"
          :validation-message="hasValidationMessage"></slot>
    <m-validation-message v-if="displayValidation"
                          class="m-input-group__validation__message"
                          :disabled="isDisabled"
                          :waiting="isWaiting"
                          :error="hasError"
                          :error-message="errorMessage"
                          :valid="isValid"
                          :valid-message="validMessage"
                          :helper-message="helperMessage"></m-validation-message>
</fieldset>
