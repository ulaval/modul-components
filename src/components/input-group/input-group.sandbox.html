<div>
    <!-- Sandbox template -->
    <h3>Tools</h3>

    <m-textfield label="Message valide"
                 v-model="validMessage"></m-textfield><br /><br />
    <m-textfield label="Message d'erreur"
                 v-model="errorMessage"></m-textfield><br /><br />
    <m-textfield label="Message d'aide"
                 v-model="helperMessage"></m-textfield><br /><br />
    <button @click="isDisabled = !isDisabled">Toogle disabled (current value: {{ !!isDisabled }}) </button><br />
    <button @click="isWaiting = !isWaiting">Toogle waiting (current value: {{ !!isWaiting }}) </button><br />
    <button @click="isReadonly = !isReadonly">Toogle readonly (current value: {{ !!isReadonly }}) </button><br />
    <button @click="isValid = !isValid">Toogle valid (current value: {{ !!isValid }}) </button><br />
    <button @click="hasError = !hasError">Toogle error (current value: {{ !!hasError }}) </button><br />
    <button @click="visible = !visible">Toogle fieldset visiblity (current value: {{ !!visible }}) </button><br />


    <h3>Propriétés ré-exposées par le composant</h3>
    <p class="m-u--margin-bottom">
        Les propriétés ré-exposées peuvent être assignées aux objets passés par la slot
    </p>
    <p>
        <m-input-group legend="Propriétés exposées"
                       :visible="visible"
                       :display-validation="false"
                       :error-message="errorMessage"
                       :error="hasError"
                       :valid-message="validMessage"
                       :valid="isValid"
                       :helper-message="helperMessage"
                       :disabled="isDisabled"
                       :waiting="isWaiting"
                       :readonly="isReadonly">
            <div slot-scope="{ disabled, waiting, readonly, valid, error, validationMessage }">
                <dl>
                    <dt>disabled</dt>
                    <dd :class="{ stronger : disabled }">{{ disabled }}</dd>
                    <dt>waiting</dt>
                    <dd :class="{ stronger : waiting }">{{ waiting }}</dd>
                    <dt>readonly</dt>
                    <dd :class="{ stronger : readonly }">{{ readonly }}</dd>
                    <dt>valid</dt>
                    <dd :class="{ stronger : valid }">{{ valid }} -- Cette propriété réagit au message valide</dd>
                    <dt>error</dt>
                    <dd :class="{ stronger : error }">{{ error }} -- Cette propriété réagit au message d'erreur</dd>
                    <dt>validationMessage</dt>
                    <dd :class="{ stronger : validationMessage }">{{ validationMessage }} -- Cette propriété réagit aux messages (help, valid, error)</dd>
                </dl>
            </div>
        </m-input-group>
    </p>
    <h3>Utilisation avec des champs textes (m-textfield)</h3>
    <p class="m-u--margin-bottom">
        Un exemple de validation de données est inclut
    </p>
    <p>
        <m-input-group legend="Gestion de champs de texte"
                       :visible="visible"
                       :error-message="errorMessage"
                       :error="hasError"
                       :valid-message="validMessage"
                       :valid="isValid"
                       :helper-message="helperMessage"
                       :disabled="isDisabled"
                       :waiting="isWaiting"
                       :readonly="isReadonly">
            <div slot-scope="{ disabled, waiting, readonly, valid, error }">
                <m-textfield label="Prénom"
                             :disabled="disabled"
                             :waiting="waiting"
                             :readonly="readonly"
                             :error="error"
                             :valid="valid"
                             v-model="prenom"
                             @input="valider"></m-textfield>
                <m-textfield label="initiale"
                             :max-width="xSmallFormWidth"
                             :disabled="disabled"
                             :waiting="waiting"
                             :readonly="readonly"
                             :error="error"
                             :valid="valid"
                             v-model="initiale"></m-textfield>
                <m-textfield label="Nom de famille"
                             :disabled="disabled"
                             :waiting="waiting"
                             :readonly="readonly"
                             :error="error"
                             :valid="valid"
                             v-model="nom"
                             @input="valider"></m-textfield>
            </div>
        </m-input-group>
    </p>
    <h3>Utilisation avec des champs textes (m-textfield) en ignorant les messages de validation regroupés</h3>
    <p class="m-u--margin-bottom">
        Un exemple de validation de données est inclut
    </p>
    <p>
        <m-input-group legend="Gestion de champs de texte"
                       :visible="visible"
                       :error="hasError"
                       :valid="isValid"
                       :disabled="isDisabled"
                       :waiting="isWaiting"
                       :readonly="isReadonly">
            <div slot-scope="{ disabled, waiting, readonly, valid, error }">
                <m-textfield label="Prénom"
                             :disabled="disabled"
                             :waiting="waiting"
                             :readonly="readonly"
                             :error="error"
                             :valid="valid"
                             :error-message="errorMessagePrenom"
                             :valid-message="validMessage"
                             :helper-message="helperMessage"
                             v-model="prenom"
                             @input="valider"></m-textfield>
                <m-textfield label="initiale"
                             :max-width="xSmallFormWidth"
                             :disabled="disabled"
                             :waiting="waiting"
                             :readonly="readonly"
                             :error="error"
                             :valid="valid"
                             :valid-message="validMessage"
                             :helper-message="helperMessage"
                             v-model="initiale"></m-textfield>
                <m-textfield label="Nom de famille"
                             :disabled="disabled"
                             :waiting="waiting"
                             :readonly="readonly"
                             :error="error"
                             :valid="valid"
                             :error-message="errorMessageNom"
                             :valid-message="validMessage"
                             :helper-message="helperMessage"
                             v-model="nom"
                             @input="valider"></m-textfield>
            </div>
        </m-input-group>
    </p>
    <h3>Utilisation avec des champs variés (m-textfield)</h3>
    <p>
        <m-input-group legend="Gestion de champs de texte"
                       :visible="visible"
                       :error-message="errorMessage"
                       :error="hasError"
                       :valid-message="validMessage"
                       :valid="isValid"
                       :helper-message="helperMessage"
                       :disabled="isDisabled"
                       :waiting="isWaiting"
                       :readonly="isReadonly">
            <div slot-scope="{ disabled, waiting, readonly, valid, error }">
                <m-rich-text-editor v-model="texteRiche"
                                    :label="'texte riche'"
                                    :max-width="'large'"
                                    :placeholder="'some placeholder'"
                                    :error="error"
                                    :valid="valid"
                                    :waiting="waiting"
                                    :disabled="disabled"
                                    :readonly="readonly"></m-rich-text-editor>
                <hr />
                <m-radio-group :error="error"
                               :valid="valid"
                               :read-only="readonly"
                               :disabled="disabled || waiting">
                    <m-radio value="1">1</m-radio>
                    <m-radio value="2">2</m-radio>
                    <m-radio value="3">3</m-radio>
                </m-radio-group>
                <hr />
                <m-dropdown label="Légume"
                            v-model="legumes"
                            :error="error"
                            :valid="valid"
                            :waiting="waiting"
                            :disabled="disabled"
                            :readonly="readonly">
                    <m-dropdown-item value="6a"
                                     label="Asperge"></m-dropdown-item>
                    <m-dropdown-item value="6b"
                                     label="Brocoli"></m-dropdown-item>
                    <m-dropdown-item value="6c"
                                     label="Carotte"></m-dropdown-item>
                    <m-dropdown-item value="6d"
                                     label="Daikon"></m-dropdown-item>
                    <m-dropdown-item value="6e"
                                     label="Épinard"></m-dropdown-item>
                </m-dropdown>
            </div>
        </m-input-group>
    </p>
</div>
