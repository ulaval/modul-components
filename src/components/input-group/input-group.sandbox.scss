.stronger {
    font-weight: 900;
}

hr {
    border: 1px solid hotpink;
}
