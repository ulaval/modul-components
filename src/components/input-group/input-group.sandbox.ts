import Vue, { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { InputMaxWidth } from '../../mixins/input-width/input-width';
import { INPUT_GROUP_NAME } from '../component-names';
import { MRichTextEditor } from '../rich-text-editor/rich-text-editor';
import TextfieldPlugin from '../textfield/textfield';
import InputGroupPlugin from './input-group';
import WithRender from './input-group.sandbox.html?style=./input-group.sandbox.scss';


@WithRender
@Component({
    components: { MRichTextEditor }
})
export class MInputGroupSandbox extends Vue {

    validMessage: string = '';
    errorMessage: string = '';
    helperMessage: string = '';

    isDisabled: boolean = false;
    isReadonly: boolean = false;
    isWaiting: boolean = false;
    isValid: boolean = false;
    hasError: boolean = false;
    visible: boolean = true;

    nom: string = '';
    prenom: string = '';
    initiale: string = '';

    texteRiche: string = '';
    legumes: string = '';

    get xSmallFormWidth(): string {
        return InputMaxWidth.XSmall;
    }

    get errorMessageNom(): string {
        if (!this.nom) {
            return 'Le nom est obligatoire';
        }
        return '';
    }

    get errorMessagePrenom(): string {
        if (!this.prenom) {
            return 'Le pr√©nom est obligatoire';
        }
        return '';
    }

    valider(): void {
        this.errorMessage = '';
        this.validerNom();
        this.validerPrenom();
    }

    validerNom(): void {
        this.errorMessage = (this.errorMessageNom) ? this.errorMessageNom : this.errorMessage;
    }

    validerPrenom(): void {
        this.errorMessage = (this.errorMessagePrenom) ? this.errorMessagePrenom : this.errorMessage;
    }
}

const InputGroupSandboxPlugin: PluginObject<any> = {
    install(v, options): void {
        v.component(`${INPUT_GROUP_NAME}-sandbox`, MInputGroupSandbox);
        v.use(InputGroupPlugin);
        v.use(TextfieldPlugin);
    }
};

export default InputGroupSandboxPlugin;
