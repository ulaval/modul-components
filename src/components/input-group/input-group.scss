@import 'commons';
@import './../input-style/input-style.scss';

.m-input-group {
    border: 1px solid $m-color--grey-light;
    border-radius: $m-border-radius--s;

    &.m--border-hidden {
        border-width: 0;
    }

    &__legend {
        color: $m-color--text-light;
    }

    &.m--is-readonly,
    &.m--is-disabled {
        border-style: dashed;

        .m-input-group__legend {
            color: $m-input-style--color-disabled;
        }
    }

    &.m--is-waiting {
        @include input-style-cursor(wait);
    }

    &.m--is-valid {
        border-color: $m-input-style--color-valid;

        .m-input-group__legend {
            color: $m-input-style--color-valid;
        }
    }

    &.m--has-error {
        border-color: $m-input-style--color-error;

        .m-input-group__legend {
            color: $m-input-style--color-error;
        }
    }

    &__validation__message {
        margin-top: $m-spacing--xs;
    }
}
