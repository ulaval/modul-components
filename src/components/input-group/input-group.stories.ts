import { storiesOf } from '@storybook/vue';
import Vue from 'vue';
import { componentsHierarchyRootSeparator } from '../../../conf/storybook/utils';
import { INPUT_GROUP_NAME } from '../component-names';
import InputGroupPlugin from './input-group';

Vue.use(InputGroupPlugin);

storiesOf(`${componentsHierarchyRootSeparator}${INPUT_GROUP_NAME}`, module)

    .add('default', () => ({
        template: `<div>
                        <m-input-group>
                            <m-checkbox>Checkbox 1</m-checkbox>
                            <m-checkbox>Checkbox 2</m-checkbox>
                            <m-checkbox>Checkbox 3</m-checkbox>
                        </m-input-group>
                        <m-input-group class="m-u--margin-top--l">
                            <m-textfield label="Texfield 1"></m-textfield>
                            <m-textfield label="Texfield 2"></m-textfield>
                        </m-input-group>
                    </div>`
    }))
    .add('label', () => ({
        template: `<m-input-group label="This is a label">
                       <m-checkbox>Checkbox 1</m-checkbox>
                       <m-checkbox>Checkbox 2</m-checkbox>
                       <m-checkbox>Checkbox 3</m-checkbox>
                   </m-input-group>`
    }))
    .add('required-marker', () => ({
        template: `<m-input-group label="This is a label" :required-marker="true">
                       <m-checkbox>Checkbox 1</m-checkbox>
                       <m-checkbox>Checkbox 2</m-checkbox>
                       <m-checkbox>Checkbox 3</m-checkbox>
                   </m-input-group>`
    }))
    .add('icon slot', () => ({
        template: `<m-input-group label="This is a label" :required-marker="true">
                       <template #icon>
                           <m-icon name="m-svg__add-circle" />
                       </template>
                       <m-checkbox>Checkbox 1</m-checkbox>
                       <m-checkbox>Checkbox 2</m-checkbox>
                       <m-checkbox>Checkbox 3</m-checkbox>
                   </m-input-group>`
    }))
    .add('error-message', () => ({
        template: `<m-input-group label="This is a label" error-message="This is an error message">
                       <m-checkbox>Checkbox 1</m-checkbox>
                       <m-checkbox>Checkbox 2</m-checkbox>
                       <m-checkbox>Checkbox 3</m-checkbox>
                   </m-input-group>`
    }))
    .add('valid-message', () => ({
        template: `<m-input-group label="This is a label" valid-message="This is an valid message">
                       <m-checkbox>Checkbox 1</m-checkbox>
                       <m-checkbox>Checkbox 2</m-checkbox>
                       <m-checkbox>Checkbox 3</m-checkbox>
                   </m-input-group>`
    }))
    .add('helper-message', () => ({
        template: `<m-input-group label="This is a label" helper-message="This is an helper message">
                       <m-checkbox>Checkbox 1</m-checkbox>
                       <m-checkbox>Checkbox 2</m-checkbox>
                       <m-checkbox>Checkbox 3</m-checkbox>
                   </m-input-group>`
    }));

storiesOf(`${componentsHierarchyRootSeparator}${INPUT_GROUP_NAME}/validation-message-position`, module)
    .add('top', () => ({
        template: `<m-input-group label="This is a label" error-message="This is an error message" validation-message-position="top">
                       <m-checkbox>Checkbox 1</m-checkbox>
                       <m-checkbox>Checkbox 2</m-checkbox>
                       <m-checkbox>Checkbox 3</m-checkbox>
                   </m-input-group>`
    }))
    .add('bottom', () => ({
        template: `<m-input-group label="This is a label" error-message="This is an error message" validation-message-position="bottom">
                       <m-checkbox>Checkbox 1</m-checkbox>
                       <m-checkbox>Checkbox 2</m-checkbox>
                       <m-checkbox>Checkbox 3</m-checkbox>
                   </m-input-group>`
    }))
    .add('top - no label', () => ({
        template: `<m-input-group error-message="This is an error message" validation-message-position="top">
                       <m-checkbox>Checkbox 1</m-checkbox>
                       <m-checkbox>Checkbox 2</m-checkbox>
                       <m-checkbox>Checkbox 3</m-checkbox>
                   </m-input-group>`
    }))
    .add('bottom - no label', () => ({
        template: `<m-input-group error-message="This is an error message" validation-message-position="bottom">
                       <m-checkbox>Checkbox 1</m-checkbox>
                       <m-checkbox>Checkbox 2</m-checkbox>
                       <m-checkbox>Checkbox 3</m-checkbox>
                   </m-input-group>`
    }));
