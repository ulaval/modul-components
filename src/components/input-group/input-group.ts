import { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import uuid from '../../utils/uuid/uuid';
import { INPUT_GROUP_NAME } from '../component-names';
import ValidationMessagePlugin from '../validation-message/validation-message';
import { InputManagement } from './../../mixins/input-management/input-management';
import { InputState } from './../../mixins/input-state/input-state';
import { ModulVue } from './../../utils/vue/vue';
import WithRender from './input-group.html?style=./input-group.scss';

export enum MMInputGroupValidationMessagePosition {
    Top = 'top',
    Bottom = 'bottom'
}

export interface MInputGroupProps {
    readonly label: string;
    readonly requiredMarker?: boolean;
    readonly validationMessagePosition?: MMInputGroupValidationMessagePosition;
}

@WithRender
@Component({
    mixins: [
        InputState,
        InputManagement
    ]
})
export class MInputGroup extends ModulVue implements MInputGroupProps {

    @Prop()
    readonly label!: string;
    @Prop()
    readonly requiredMarker!: boolean;

    @Prop({
        default: MMInputGroupValidationMessagePosition.Top,
        validator: value =>
            value === MMInputGroupValidationMessagePosition.Top ||
            value === MMInputGroupValidationMessagePosition.Bottom
    })
    public validationMessagePosition: MMInputGroupValidationMessagePosition;

    get hasLabel(): boolean {
        return !!this.label;
    }

    get isValidationMessagePositionBottom(): boolean {
        return this.validationMessagePosition === MMInputGroupValidationMessagePosition.Bottom;
    }

    get idLabel(): string | undefined {
        return this.hasLabel ? uuid.generate() : undefined;
    }

    get idValidationMessage(): string | undefined {
        return this.as<InputState>().errorMessage || this.as<InputState>().validMessage || this.as<InputState>().helperMessage ? uuid.generate() : undefined;
    }
}


const InputGroupPlugin: PluginObject<any> = {
    install(v, options): void {
        v.use(ValidationMessagePlugin);
        v.component(INPUT_GROUP_NAME, MInputGroup);
    }
};

export default InputGroupPlugin;
