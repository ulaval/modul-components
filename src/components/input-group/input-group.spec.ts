import { mount, Wrapper } from '@vue/test-utils';
import Vue from 'vue';
import { renderComponent } from '../../../tests/helpers/render';
import { MInputGroup } from '../input-group/input-group';

describe(`MInputGroup`, () => {
    let wrapper: Wrapper<MInputGroup>;
    let legend: string;
    let visible: boolean;
    let disabled: boolean;
    let waiting: boolean;
    let readonly: boolean;
    let valid: boolean;
    let error: boolean;
    let errorMessage: string;
    let validMessage: string;
    let helperMessage: string;
    let displayValidation: boolean;

    beforeEach(() => {
        legend = undefined as any;
        visible = undefined as any;
        disabled = undefined as any;
        waiting = undefined as any;
        readonly = undefined as any;
        valid = undefined as any;
        error = undefined as any;
        errorMessage = undefined as any;
        validMessage = undefined as any;
        helperMessage = undefined as any;
        displayValidation = undefined as any;

    });

    const initialiserWrapper: Function = (): void => {
        wrapper = mount(MInputGroup, {
            localVue: Vue,
            propsData: {
                legend,
                visible,
                disabled,
                waiting,
                readonly,
                valid,
                error,
                errorMessage,
                validMessage,
                helperMessage,
                displayValidation
            },
            scopedSlots: {
                default: '<div slot-scope="result">{{ result.disabled }} {{ result.error }} {{ result.valid }} {{ result.readonly }} {{ result.validationMessage }}</div>'
            }
        });
    };

    describe(`with default values`, () => {
        beforeEach(() => {
            initialiserWrapper();
        });

        it(`will render correctly`, async () => {
            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });

    });

    describe(`with all props defined`, () => {
        beforeEach(() => {
            legend = 'Some legend';
            visible = true;
            disabled = true;
            waiting = true;
            readonly = true;
            valid = true;
            error = true;
            errorMessage = 'Some error message';
            validMessage = 'Some valid message';
            helperMessage = 'Some helper message';
            displayValidation = true;
            initialiserWrapper();
        });

        it(`will render correctly`, async () => {
            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });

        describe(`when setting visiblity to false`, () => {
            beforeEach(() => {
                visible = false;
                displayValidation = false;
                initialiserWrapper();
            });

            it(`will render correctly`, async () => {
                expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
            });
        });

        describe(`when removing legend`, () => {
            beforeEach(() => {
                legend = '';
                initialiserWrapper();
            });

            it(`will render correctly`, async () => {
                expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
            });
        });
    });
});
