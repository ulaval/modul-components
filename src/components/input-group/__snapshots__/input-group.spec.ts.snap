// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MInputGroup with all props defined when removing legend will render correctly 1`] = `
<fieldset class="m-input-group m--is-readonly m--is-disabled">
  <div>true false false true false</div>
  <m-validation-message disabled="disabled" error-message="Some error message" valid-message="Some valid message" helper-message="Some helper message" class="m-input-group__validation__message"></m-validation-message>
</fieldset>
`;

exports[`MInputGroup with all props defined when setting visiblity to false will render correctly 1`] = `
<fieldset class="m-input-group m--is-readonly m--is-disabled m--border-hidden">
  <div>true false false true false</div>
</fieldset>
`;

exports[`MInputGroup with all props defined will render correctly 1`] = `
<fieldset class="m-input-group m--is-readonly m--is-disabled">
  <legend class="m-input-group__legend">Some legend</legend>
  <div>true false false true false</div>
  <m-validation-message disabled="disabled" error-message="Some error message" valid-message="Some valid message" helper-message="Some helper message" class="m-input-group__validation__message"></m-validation-message>
</fieldset>
`;

exports[`MInputGroup with default values will render correctly 1`] = `
<fieldset class="m-input-group">
  <div>false false false false</div>
  <m-validation-message class="m-input-group__validation__message"></m-validation-message>
</fieldset>
`;
